runtime/nsd_ossl.c

```diff
@@ -215,22 +219,595 @@ void osslLastSSLErrorMsg(int ret, SSL *ssl, int severity, const char* pszCallSou
+			goto err;
+		} else {
+			LogError(0, RS_RET_NO_ERRCODE, "OCSP: Certificate \"%s\" has been revoked: "
+					"No reason provided.\n", cert_name);
+			goto err;
+		}
+		break;
+	case V_OCSP_CERTSTATUS_UNKNOWN:
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Status of certificate \"%s\" is unknown. "
+				"(This is a fatal error).\n", cert_name);
+		goto err;
+		break;
+	default:
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Unexpected certificate status \"%d\" for %s .\n",
+				status, cert_name);
+		goto err;
+		break;
+	}
+
+
+	ret = 1;
+err:
+
+	if (bs)
+		OCSP_BASICRESP_free(bs);
+
+	return ret;
+}
+
+static BIO* ocsp_connect(const char *host, const char *port, const char *device)
+{
+	int erropt, ret = 0, fd = 0;
+	struct timeval tv;
+	fd_set confds;
+	struct addrinfo hints;
+	struct addrinfo *rp = NULL;
+	BIO *bio = NULL;
+	socklen_t errlen;
+
+	memset(&hints, 0, sizeof(struct addrinfo));
+	hints.ai_family = AF_UNSPEC;
+	hints.ai_socktype = SOCK_STREAM;
+
+	ret = getaddrinfo(host, port, &hints, &rp);
+	if (ret != 0) {
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Could not lookup address for %s:%s\n", host, port);
+		goto err;
+	}
+
+	fd = socket(rp->ai_family, SOCK_STREAM, 0);
+	if (fd == -1) {
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Could not create socket for %s:%s\n", host, port);
+		goto err;
+	}
+
+	if (device) {
+#if defined(SO_BINDTODEVICE)
+		if(setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, device,
+		              strlen(device) + 1) < 0)
+#endif
+		{
+			LogError(0, RS_RET_NO_ERRCODE, "Failed to bind to interface \"%s\" for "
+					"OCSP request.\n", device);
+			goto err;
+		}
+	}
+	if (fcntl(fd, F_SETFL, O_NONBLOCK) == -1) {
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Failed make socket non-blocking for %s:%s: %s\n",
+				host, port, strerror(errno));
+		goto err;
+	}
+
+	if (connect(fd, rp->ai_addr, rp->ai_addrlen) != -1) {
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Failed to connect to %s:%s: %s\n",
+				host, port, strerror(errno));
+		goto err;
+	}
+
+	FD_ZERO(&confds);
+	FD_SET(fd, &confds);
+	tv.tv_usec = 0;
+	tv.tv_sec = OCSP_TIMEOUT;
+	ret = select(fd + 1, NULL, (void *)&confds, NULL, &tv);
+	if (ret == 0) {
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Connection timeout while connecting to %s:%s\n",
+				host, port);
+		goto err;
+	}
+
+	errlen = sizeof(erropt);
+	if (getsockopt(fd, SOL_SOCKET, SO_ERROR, &erropt, &errlen) != 0) {
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Failed to socket status for connection %s:%s: %s\n",
+				host, port, strerror(errno));
+		goto err;
+	}
+
+	if ((erropt != 0) && (erropt != EISCONN)) {
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Failed to establish connection for %s:%s: %s\n",
+				host, port, strerror(erropt));
+		goto err;
+	}
+
+	bio = BIO_new_socket(fd, BIO_CLOSE);
+	if (bio == NULL) {
+		LogError(0, RS_RET_NO_ERRCODE, "OCSP: Could attach socket to BIO object\n");
+		goto err;
+	}
+err:
+
+	freeaddrinfo(rp);
+
+	if (!bio) /* don't close the socket in case of success */
```

Comment:
try to change it to 

if (bio != NULL)

to pass codacy.

Reply:
Long time no see Daniel ;)

> @rgerhards , @alorbach , is there any trick to make Codacy happy with respect to the "unknown" SO_BINDTODEVICE value?

I think this issue is a false positive and can be ignored. 
I have commented above on the other issue.

Reply:
Indeed, pretty long time :)

Pushed a bio==NULL variant.

Let's see if that passes codacy.


Do you have any preference or suggestion on additional configuration knobs?

I guess a knob to turn of all certificate revocation checks off would probably be helpful, to mimic the previous behavior. Personally I would prefer if (online) revocation checks would be done by default, which might be expected by some TLS users.

I could contribute online CRL support as well, but that is rather complex, with delta-CRLs and such things. And not sure if online CRLs has really a future. Most or all public CAs do OCSP today, might be also a CAB Forum requirement (haven't checked). Letsencrypt does OCSP-only. So I wonder if online CRL is worth the burden.

Reply:
No luck with codacy:

   `Condition 'bio==NULL' is always false`

Reply:
> No luck with codacy:
> 
> `Condition 'bio==NULL' is always false`

@rgerhards do you see the reason why this is detected as issue?

Reply:
> Indeed, pretty long time :)
> 
> Pushed a bio==NULL variant.
> 
> Let's see if that passes codacy.
> 
> Do you have any preference or suggestion on additional configuration knobs?

Yes indeed I would recommend to have configuration options in imtcp and omfwd enabling  / disabling OCSP checking with default OFF to maintain existing behavior. 
With a name like "StreamDriver.TlsVerifyOcsp" (See other TLS Options in imtcp/omfwd doc): 
https://www.rsyslog.com/doc/master/configuration/modules/imtcp.html
https://www.rsyslog.com/doc/master/configuration/modules/omfwd.html

You needs some additional coding to pass the configuration option from the module down to the nsd driver. And not implented wrappers in nsd_gtls and nsd_ptcp drivers. Here is an older PR where I added "streamdriver.permitexpiredcerts" which you can use a template: 
https://github.com/rsyslog/rsyslog/pull/3366

> I could contribute online CRL support as well, but that is rather complex, with delta-CRLs and such things. And not sure if online CRLs has really a future. Most or all public CAs do OCSP today, might be also a CAB Forum requirement (haven't checked). Letsencrypt does OCSP-only. So I wonder if online CRL is worth the burden.

For now I would also recommend too leave it with OCSP for now. If CRL becomes more important in future, it can be discussed again.

Btw., it would be great if we could get a testcase for OCSP. Do you have an idea how to do make one that works with a free and reliable OCSP service?

Reply:
> > Indeed, pretty long time :)
> > Pushed a bio==NULL variant.
> > Let's see if that passes codacy.
> > Do you have any preference or suggestion on additional configuration knobs?
> 
> Yes indeed I would recommend to have configuration options in imtcp and omfwd enabling / disabling OCSP checking with default OFF to maintain existing behavior.
> With a name like "StreamDriver.TlsVerifyOcsp" (See other TLS Options in imtcp/omfwd doc):
> https://www.rsyslog.com/doc/master/configuration/modules/imtcp.html
> https://www.rsyslog.com/doc/master/configuration/modules/omfwd.html
> 
> You needs some additional coding to pass the configuration option from the module down to the nsd driver. And also "not implented" wrappers in nsd_gtls and nsd_ptcp drivers. Here is an older PR where I added "streamdriver.permitexpiredcerts" which you can use a template:
> #3366


Ok, I'll look into introduce those knobs.

With respect to the default behavior: Wouldn't a user/operator expect that the TLS drivers perform such online revocation check by default, if the CA/PKI is including those? E.g. if someone hosts a TLS rsyslog server with a public CA vendor, the server gets compromised the  end-entity certificate of the server revoked by the CA. The OCSP server would know about the revoked certificate, the TLS rsyslog clients could then be "protected" talking to a compromised/revoked certificate. At least that would be my naive expectation that TLS client comply with: https://tools.ietf.org/html/rfc5280#section-6 /  https://tools.ietf.org/html/rfc5280#section-6.3
(which is referred by https://tools.ietf.org/html/rfc5425#section-4.2.1 )


I understand that this might break existing deployments under certain circumstances. But on the long run I would encourage to perform online revocation checks if OCSP (or CRL) URL are embedded in the  end-entity or in the certificate path:

```
$ openssl s_client -connect github.com:443 < /dev/null 2> /dev/null | openssl x509 -text -noout | grep -A5 -E '(Authority Information Access|CRL Distribution Points)'
            X509v3 CRL Distribution Points: 

                Full Name:
                  URI:http://crl3.digicert.com/sha2-ha-server-g6.crl

                Full Name:
--
            Authority Information Access: 
                OCSP - URI:http://ocsp.digicert.com
                CA Issuers - URI:http://cacerts.digicert.com/DigiCertSHA2HighAssuranceServerCA.crt

            X509v3 Basic Constraints: critical
                CA:FALSE
```

This would not affect existing deployment using private PKI environment without any CRL/OCSP URL in the issued certificate.

Not sure if "ignoring" OCSP/CRL URLs as part of certificate validation would be categorized as https://cwe.mitre.org/data/definitions/299.html

(Yes, the browsers don't do by default live online reovcation either IIUC, but they have some sort of regular caches updated transparently to the user. So there is at least some revocation checking going on)

On the long run OCSP stapling support would be obviously nice. The initial motivation for us implementing this was fulfilling some security requirements as syslog+TLS client, not as server. Not sure if there is on the market already a OCSP stapling aware syslog server. Maybe something worth to look into to add for rsyslog in future.

> 
> > I could contribute online CRL support as well, but that is rather complex, with delta-CRLs and such things. And not sure if online CRLs has really a future. Most or all public CAs do OCSP today, might be also a CAB Forum requirement (haven't checked). Letsencrypt does OCSP-only. So I wonder if online CRL is worth the burden.
> 
> For now I would also recommend too leave it with OCSP for now. If CRL becomes more important in future, it can be discussed again.
> 
> Btw., it would be great if we could get a testcase for OCSP. Do you have an idea how to do make one that works with a free and reliable OCSP service?

Sure, first you need to issue a set of certificates with a OCSP URI (Authority Information Access) included. To generate such testing purposes I build this: https://github.com/danos/vyatta-security-vpn/tree/master/docs/x509

I'll look into the existing rsyslog UTs if I can enhance/add something to existing ossl related UTs.

For development I used openssl as OCSP responder/server:
```
openssl ocsp -index index.txt -port 8888 -rsigner RSA_ocsp1.pem -rkey RSA_ocsp1.key -CA RSA_CA.pem -text -nmin 10
```

The index.txt is the file which openssl generates when you use OpenSSL to build a PKI. In this example the certificate for server13 got revoked. And `openssl ocsp -index indext.txt` basically uses that file as revocation database.

```
$ cat index.txt
V       191110011802Z           00      unknown /C=US/ST=CA/O=AT&T/OU=Labs/CN=client1/emailAddress=client1+rsa@vyatta.att-mail.com
V       191110011816Z           01      unknown /C=US/ST=CA/O=AT&T/OU=Labs/CN=serverA/emailAddress=serverA+rsa@vyatta.att-mail.com
R       191110011822Z   181110011831Z   02      unknown /C=US/ST=CA/O=AT&T/OU=Labs/CN=server13/emailAddress=server13+rsa@vyatta.att-mail.com
```

To make use of OCSP nonces add `-nonce`:

```
openssl ocsp -index index.txt -port 8888 -rsigner RSA_ocsp1.pem -rkey RSA_ocsp1.key -CA RSA_CA.pem -text -nmin 10
```


I haven't checked other free/open implementations yet. FreeIPA/dogtag PKI might be worth a look. But for self-contained UTs `openssl ocsp` might be sufficient I guess.

The only other detailed behavior change between different production OCSP responder/server deployments, which are easy to overlook if you just use `openssl ocsp`: the HTTP requests need to hold a "Host" header entry. Most or many public CA vendor seem to use CDN providers to guarantee their OCSP availability and responsiveness.

Reply:
> With respect to the default behavior: Wouldn't a user/operator expect that the TLS drivers perform such online revocation check by default, if the CA/PKI is including those? E.g. if someone hosts a TLS rsyslog server with a public CA vendor, the server gets compromised the end-entity certificate of the server revoked by the CA. The OCSP server would know about the revoked certificate, the TLS rsyslog clients could then be "protected" talking to a compromised/revoked certificate. At least that would be my naive expectation that TLS client comply with: https://tools.ietf.org/html/rfc5280#section-6 / https://tools.ietf.org/html/rfc5280#section-6.3
> (which is referred by https://tools.ietf.org/html/rfc5425#section-4.2.1 )
> 
> I understand that this might break existing deployments under certain circumstances. But on the long run I would encourage to perform online revocation checks if OCSP (or CRL) URL are embedded in the end-entity or in the certificate path:

Thanks for the details, @rgerhards I think under these circumstances it is save and reasonable to have OCSP checking enabled by default with the possibility to disable it if needed.

> On the long run OCSP stapling support would be obviously nice. The initial motivation for us implementing this was fulfilling some security requirements as syslog+TLS client, not as server. Not sure if there is on the market already a OCSP stapling aware syslog server.

Probably not yet ;).

> Sure, first you need to issue a set of certificates with a OCSP URI (Authority Information Access) included. To generate such testing purposes I build this: https://github.com/danos/vyatta-security-vpn/tree/master/docs/x509
> 
> I'll look into the existing rsyslog UTs if I can enhance/add something to existing ossl related UTs.
> 
> For development I used openssl as OCSP responder/server:
> 
> ```
> openssl ocsp -index index.txt -port 8888 -rsigner RSA_ocsp1.pem -rkey RSA_ocsp1.key -CA RSA_CA.pem -text -nmin 10
> ```
> 
> The index.txt is the file which openssl generates when you use OpenSSL to build a PKI. In this example the certificate for server13 got revoked. And `openssl ocsp -index indext.txt` basically uses that file as revocation database.
> 
> ```
> $ cat index.txt
> V       191110011802Z           00      unknown /C=US/ST=CA/O=AT&T/OU=Labs/CN=client1/emailAddress=client1+rsa@vyatta.att-mail.com
> V       191110011816Z           01      unknown /C=US/ST=CA/O=AT&T/OU=Labs/CN=serverA/emailAddress=serverA+rsa@vyatta.att-mail.com
> R       191110011822Z   181110011831Z   02      unknown /C=US/ST=CA/O=AT&T/OU=Labs/CN=server13/emailAddress=server13+rsa@vyatta.att-mail.com
> ```
> 
> To make use of OCSP nonces add `-nonce`:
> 
> ```
> openssl ocsp -index index.txt -port 8888 -rsigner RSA_ocsp1.pem -rkey RSA_ocsp1.key -CA RSA_CA.pem -text -nmin 10 -nonce
> ```
That would be fully sufficient as ocsp server to build a testcase on. 
It just needs to be started in background and killed on shutdown. 
This needs some additional plumbing in diag.sh like done for this test here:
https://github.com/rsyslog/rsyslog/blob/master/tests/sndrcv_omsnmpv1_udp.sh
(See snmp_start_trapreceiver  and snmp_stop_trapreceiver )

If you create sample ca / certifcates based on your script, but make them last longer (Like 36500 days) which is ok for the testbench, add them to here for example (tests/testsuites/x.509), you can use these tests as template to build your own test on: 

https://github.com/rsyslog/rsyslog/blob/master/tests/sndrcv_tls_certvalid.sh
(Just replace '$RS_TLS_DRIVER' with ossl, as this will only run for openssl). 
And this tests: 
https://github.com/rsyslog/rsyslog/blob/master/tests/imtcp-tls-ossl-basic.sh

You can run single tests manually when you build rsyslog like this:
- go to rsyslog source dir
- export srcdir=./
- autoreconf -vfi 
- ./configure ... your options ...
- make && make check 
- ...wait until first tests runs then break with CTRL + C
- cd tests
- ./yourtestcase.sh



> I haven't checked other free/open implementations yet. FreeIPA/dogtag PKI might be worth a look. But for self-contained UTs `openssl ocsp` might be sufficient I guess.
> 
> The only other detailed behavior change between different production OCSP responder/server deployments, which are easy to overlook if you just use `openssl ocsp`: the HTTP requests need to hold a "Host" header entry. Most or many public CA vendor seem to use CDN providers to guarantee their OCSP availability and responsiveness.

If you could generate ca & test certificates using OCSP Url pointing to http://testbench.rsyslog.com/testbench/